python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+import inspect
+
+
+# A simple `partialmethod` that allows setting attributes such as
+# __doc__ on instances.
+def _partialmethod(method, *args1, **kwargs1):
+    def wrapper(self, *args2, **kwargs2):
+        return method(self, *args1, *args2, **kwargs1, **kwargs2)
+
+    return wrapper
+
+
+class Operation:
+    """Descriptor used to define operations for delegating mixins.
+
+    This class is designed to be assigned to the attributes (the delegating
+    methods) defined by the OperationMixin. This class will create the method
+    and mimic all the expected attributes for that method to appear as though
+    it was originally designed on the class. The use of the descriptor pattern
+    ensures that the method is only created the first time it is invoked, after
+    which all further calls use the callable generated on the first invocation.
```

Comment:
Trying to reason about this. I initially was worried it might be error prone to create these functions on the fly as they are requested rather than just starting with fully initialized objects. But I guess any kind of use case that checked to see if the method was there (like a `hasattr`) would just trigger it to be made at that time? :shrodinger:

In any case would it be helpful for perf at all to just create all of the valid methods at import time rather than runtime?